File: DSA/done/543 - diameterOfBinaryTree.py
# https://leetcode.com/problems/diameter-of-binary-tree/

from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def rootToLeaf(root):
            if not root:
                return 0

            nonlocal res
            left, right = rootToLeaf(root.left), rootToLeaf(root.right)
            res = max(res, left + right) 
            return max(left, right) + 1

        rootToLeaf(root)
        return res 

File: DSA/done/543 - diameterOfBinaryTree.ts
// https://leetcode.com/problems/diameter-of-binary-tree/

/**
 * Definition for a binary tree node.
 */
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

function diameterOfBinaryTree(root: TreeNode | null): number {
    let res = 0;

    const dfs = (node: TreeNode | null): number => {
        if (!node) return 0;
        
        const l = dfs(node.left), r = dfs(node.right);
        res = Math.max(res, l + r);
        return 1 + Math.max(l, r);
    }

    dfs(root);
    return res;
};